### PR TITLE
Make all health checks probing consistent

### LINKED ISSUE #96731: Dynamic readiness probe period time
**What would you like to be added**:

I would like a way to have the readiness probe period more dynamic.

**Why is this needed**:

We have a few use cases were we want extremely aggressive readiness probe times, because pod startup time is critical. In these cases, using an interval over 1s is detrimental. This includes cases where we have a service mesh and do not want to be slower than pods that do not have a sidecar, and cases where we are running integration tests and we want pods to be able to come up extremely quickly to avoid long tests.

Currently, if we add aggressive readiness probes, it impacts the pod for the entire lifetime, adding load on the pod and kubelet. Instead, it would be useful to be able to probe aggressively until we are initially ready, then back off to a more reasonable probing period. I could also see a use case for changing this from "until we are initially ready" to "whenever we are not ready" - either would work.

**What I have tried**

Adding a startup probe:
```yaml
        readinessProbe:
          httpGet:
            path: /get
            port: 80
            scheme: HTTP
          initialDelaySeconds: 0
          periodSeconds: 60
          successThreshold: 1
          timeoutSeconds: 1
        startupProbe:
          httpGet:
            path: /get
            port: 80
            scheme: HTTP
          initialDelaySeconds: 0
          periodSeconds: 1
          successThreshold: 1
          timeoutSeconds: 1
```

This doesn't impact things, pod still takes 60s to be marked ready

Artificial initial delay:
```
        readinessProbe:
          failureThreshold: 30
          httpGet:
            path: /get
            port: 80
            scheme: HTTP
          initialDelaySeconds: 4
          periodSeconds: 60
          successThreshold: 1
          timeoutSeconds: 1
```

In this case, we are always too fast or too slow: if we get ready in under 4s, then we wasted (4-ready time) seconds to get ready. If we happen to take 4.1 seconds to get ready, we miss our window and are stuck not ready for 60 more seconds.

/sig node

### LINKED ISSUE #98007: startupProbe takes 60s before killing a failing container
Everything is described in https://github.com/kubernetes/kubernetes/issues/96731#issuecomment-759244994
Thanks @diemus for the report!

### PR DESCRIPTION
**Pull request by matthyx**
**What type of PR is this?**
/kind bug
/kind cleanup

**What this PR does / why we need it**:
Current handling of readiness/startupProbe result updates don't include an immediate sync after declaring the container as failed.
Thus, the sync is done via slower async methods.

**Which issue(s) this PR fixes**:

Fixes #96731
Fixes #98007

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Faster update the pod status reacting on probe results. Among other changes, the first readiness probe will be called faster when startup probes succeeded, which will make Pod status as ready faster.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
labels cleanup: 

/kind bug
/kind cleanup
/priority backlog
/triage accepted

Are you sure you don't want to add a release note? May be handy

File: pkg/kubelet/prober/common_test.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/prober/common_test.go:136]: is it still WIP leftovers?

File: pkg/kubelet/prober/prober_manager_test.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/prober/prober_manager_test.go:321]: still WIP?

File: test/e2e/common/container_probe.go
Comment (SergeyKanzhelev) [inline test/e2e/common/container_probe.go:421]: copying comment from the closed PR. This is a change - original test was comparing with the smaller time (`40`) than the `sleep` duration (`45`). Presumably accounting for some delays. Just want to make sure this will not introduce flakiness

File: test/e2e/common/container_probe.go
Comment (SergeyKanzhelev) [inline test/e2e/common/container_probe.go:425]: `5s` is lower than `10s` default interval for the readiness probe, this is why it's 5, correct? This test case will benefit from more extended comment on what exactly it tests. Maybe:

 ```suggestion
			framework.Failf("Pod became ready in %v, more than 5s after startupProbe succeeded. It means that the delay readiness probes were not initiated immediately after startup finished.", startedIn)
```

File: pkg/kubelet/prober/worker.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/prober/worker.go:162]: should the `manualTriggerCh` be closed here? What's the risk of keeping it not closed ever?

File: pkg/kubelet/kubelet.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet.go:1983]: it used to say "`(container unhealthy)`". Is this piece of information worth preserving?

File: pkg/kubelet/prober/prober_manager.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/prober/prober_manager.go:255]: add logging that we failed to manually trigger the update?

File: pkg/kubelet/kubelet.go
Comment (matthyx) [reply pkg/kubelet/kubelet.go:1983]: initially this code was only used when livenessProbe failed, but now that we use it to handle success of readiness and startup too, it's incorrect to report `(container unhealthy)`

File: pkg/kubelet/prober/worker.go
Comment (matthyx) [reply pkg/kubelet/prober/worker.go:162]: well, we could say the same for the `stopCh` no?

File: pkg/kubelet/prober/prober_manager.go
Comment (matthyx) [reply pkg/kubelet/prober/prober_manager.go:255]: I don't know... does it work putting a log under `default`?

Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/prober_manager.go:255]: Hard to test, but it does: https://gobyexample.com/non-blocking-channel-operations#non-blocking-channel-operations:~:text=Therefore%20the%20default%20case%20is%20selected.

File: pkg/kubelet/kubelet.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/kubelet.go:1983]: I understand, my point was that this extra message about container being unhealthy may be useful for `livenessProbe`. Maybe this can be added somehow. Or maybe add a type of a probe and status to the message?

File: pkg/kubelet/prober/worker.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/prober/worker.go:162]: yes. Not sure what the downside will be to not close it.

Comment (matthyx) [reply pkg/kubelet/prober/worker.go:162]: I leave it like that, let's see what the approvers have to say.

**Comment by matthyx** [comment]
wanna have a look @ehashman ?

File: pkg/kubelet/kubelet.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet.go:1983]: 3 arguments for the string with 2 substitutions

Comment (SergeyKanzhelev) [reply pkg/kubelet/kubelet.go:1983]: this?
```suggestion
		klog.V(4).Infof("SyncLoop %s%s: ignore irrelevant update: %#v", probe, status, update)
```

File: pkg/kubelet/kubelet.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet.go:1986]: what a values for `probe, status` as they don't need a separator?

**Comment by SergeyKanzhelev** [comment]
lgtm, except the typo in error log.

**Comment by matthyx** [comment]
Is this really backlog, I mean there is a strong need for sidecar containers, as described in https://github.com/kubernetes/kubernetes/issues/96731

**Comment by ehashman** [comment]
@matthyx you're a k8s member, you can change the priority :)

(I personally would have put this important-longterm since there are linked issues)

**Comment by matthyx** [comment]
> @matthyx you're a k8s member, you can change the priority :)

True... I even modified the code in prow checking that...
/priority important-longterm

**Comment by matthyx** [comment]
@annajung could you please flag this as 1.21?

**Comment by annajung** [comment]
/milestone v1.21

cc @erismaster

**Review by ehashman — CHANGES_REQUESTED** [review]
Having previously dug through this code and its multitude of code paths, I'm a little worried about accidental/unintentional behaviour changes from this refactor...

File: pkg/kubelet/kubelet.go
Comment (matthyx) [reply pkg/kubelet/kubelet.go:1959]: The main benefit, apart from being easier to present/understand, is to allow me to call the `HandlePodSyncs` in a consistent manner after each *Manager update.

**Comment by matthyx** [comment]
> Having previously dug through this code and its multitude of code paths, I'm a little worried about accidental/unintentional behaviour changes from this refactor...

The code is the same... I still call `SetContainerReadiness` and `SetContainerStartup` only now I call explicitly for a pod sync to ensure we don't have to wait for the periodic reconciliation.
As you can see, it dramatically speeds up pod status updates after probe results, solving a lot of issues.
It's true, however, some things could break if some users were counting on this slowness in their workflows.

**Comment by mrunalp** [comment]
Needs rebase

**Comment by matthyx** [comment]
@SergeyKanzhelev could you have a look? I had to rebase due to bazel removal... could you also push for an approval during sig-node weekly meeting, please?

**Comment by matthyx** [comment]
/assign @derekwaynecarr
Please Derek, I need your approval for the tests...

**Comment by SergeyKanzhelev** [comment]
needs rebase now... @matthyx can you pls resolve?

File: test/e2e/common/node/container_probe.go
Comment (BenTheElder) [inline test/e2e/common/node/container_probe.go:416]: nit: this is a suuuper long line. these are slightly more difficult to review on github.com. consider splitting this up.

**Review by BenTheElder — APPROVED** [review]
/approve
for `test/e2e/`

**Comment by BenTheElder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98376/pull-kubernetes-e2e-kind-ipv6/1368037038758039552

Seems to be a failure for one of the new codepaths?

**Comment by matthyx** [comment]
/hold
This test is too flacky, depending on the load...

**Comment by matthyx** [comment]
> /hold
> This test is too flacky, depending on the load... the problem is we don't directly measure the time between startupProbe success and readinessProbe success, but rather we assume the 45s delay at the beginning is enough for the busybox container to start.
> 
> This works 100% of the time on my workstation, but maybe on the loaded cluster it's another story. I will check if it's possible to improve this detection, otherwise I will just revert the e2e test modification, as especially so close to codefreeze this would have too much of an impact.

@BenTheElder I will work on that later today if you don't mind, so get some sleep and you can review it in 8-10h with your morning coffee ;-)

**Comment by matthyx** [comment]
/hold cancel
@BenTheElder ready to review!

**Review by BenTheElder — APPROVED** [review]
/lgtm
/approve
I took getting some sleep seriously 🙃

**Comment by howardjohn** [comment]
This may deserve a release note given it will impact (in a positive way) how probes work

**Comment by matthyx** [comment]
> This may deserve a release note given it will impact (in a positive way) how probes work

Is it possible to add one in the description after merge, or it won't work and I need to do it differently?

**Comment by howardjohn** [comment]
i have no clue

**Comment by annajung** [comment]
You might need to make a PR to k/sig-release release note draft https://github.com/kubernetes/sig-release/tree/master/releases/release-1.21/release-notes, but I would double-check with @wilsonehusin, 1.21 Release notes lead. 

@wilsonehusin, Could you help answer if the description change after merge would work?

**Comment by wilsonehusin** [comment]
you can modify this PR body anytime before we start collecting. we will be gathering the next batch on Tuesday / whenever `beta.1` releases 😄

**Comment by matthyx** [comment]
> you can modify this PR body anytime before we start collecting. we will be gathering the next batch on Tuesday / whenever `beta.1` releases

Thanks @wilsonehusin I have added something, please let me know if that's clear enough.

**Comment by SergeyKanzhelev** [comment]
Maybe more text like this: 

> Faster update the pod status reacting on probe results. Among other changes, the first readiness probe will be called faster when startup probes succeeded, which will make Pod status as ready faster.

**Comment by wilsonehusin** [comment]
I usually use the heuristic of "giving reader a good-enough idea of what happened without opening the PR". From glancing through the issues mentioned, do you think the following would be accurate to describe the changes?

> First readiness probe updates Pod status faster, therefore Pods may reach ready state faster than the time defined in `readinessProbe`.

**Pull request merged**
